### PR TITLE
If the problem is deleted, its providedCodes and its testcase IOs sho…

### LINKED ISSUE #274: If the problem is deleted, its providedCodes and its testcase IOs should also be deleted
If the problem is deleted, its providedCodes and its testcase IOs should also be deleted

### PR DESCRIPTION
**Pull request by timt1028**
resolve #274

### CODE REVIEW

File: Domain/Problem/src/main/java/tw/waterball/judgegirl/problem/domain/repositories/ProblemRepository.java [hunk 1]
```diff
@@ -56,11 +56,13 @@
...
+    void archiveProblem(Problem problem);
 
-    void deleteProblemById(int problemId);
+    void deleteProblem(Problem problem);
 
     void deleteAll();
 
     void saveTags(List<String> tagList);
+
+    boolean isFileExistsById(String fileId);
```
Comment (Wally5077) [inline Domain/Problem/src/main/java/tw/waterball/judgegirl/problem/domain/repositories/ProblemRepository.java:67]: boolean existsFileById(String fileId);

File: Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java [hunk 2]
```diff
@@ -167,7 +169,18 @@ public void patchProblem(int problemId, PatchProblemParams params) {
...
+    @Override
+    public boolean isFileExistsById(String fileId) {
+        try {
+            GridFSFile gridFsFile = gridFsTemplate.findOne(new Query(Criteria.where("_id").is(fileId)));
+            gridFsTemplate.getResource(requireNonNull(gridFsFile));
+            return true;
+
+        } catch (Exception e) {
+            return false;
+        }
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java:183]:
```Java
    @Override
    public boolean existsFile() {
        return ofNullable(gridFsTemplate.findOne(new Query(where("_id").is(fileId))))
                .map(gridFsTemplate::getResource)
                .map(GridFsResource::exists)
                .orElse(false);
    }
```

File: Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java [hunk 3]
```diff
@@ -177,21 +190,45 @@ public boolean problemExists(int problemId) {
...
+                List<String> providedCodesFileIds = mapToList(problem.getLanguageEnvs().values(), LanguageEnv::getProvidedCodesFileId);
+                fileIds.addAll(providedCodesFileIds);
+            }
+            if (problem.getTestcaseIOsFileId() != null) {
+                fileIds.add(problem.getTestcaseIOsFileId());
+            }
+        });
+        if(!fileIds.isEmpty()){
+            gridFsTemplate.delete(query(where("_id").in(fileIds)));
+        }
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java:231]:
```Java
        findAll().forEach(this::deleteFiles);
```

File: Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java [hunk 4]
```diff
@@ -177,21 +190,45 @@ public boolean problemExists(int problemId) {
...
+        if (problem.getLanguageEnvs() != null) {
+            var providedCodesFileIds = mapToList(problem.getLanguageEnvs().values(), LanguageEnv::getProvidedCodesFileId);
+            fileIds.addAll(providedCodesFileIds);
+        }
+        if (problem.getTestcaseIOsFileId() != null) {
+            fileIds.add(problem.getTestcaseIOsFileId());
+        }
+        if(!fileIds.isEmpty()){
+            gridFsTemplate.delete(query(where("_id").in(fileIds)));
+        }
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java:212]: 1. List fileIds = new LinkedList<>();
2. problem.getLanguageEnvs() 和  problem.getTestcaseIOsFileId() 宣告成一個變數, 型態用 var
3. 這段抽成一個函數 deleteFiles(Problem);

File: Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java [hunk 5]
```diff
@@ -177,21 +190,45 @@ public boolean problemExists(int problemId) {
     }
 
     @Override
-    public void archiveProblemById(int problemId) {
-        Update update = new Update();
-        Query query = new Query(where("_id").is(problemId));
-        update.set("archived", true);
+    public void archiveProblem(Problem problem) {
+        Update update = Update.update("archived", true);
+        Query query = new Query(where("_id").is(problem.getId()));
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java:195]: Query query = query(where("_id").is(problem.getId()))

File: Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java [hunk 6]
```diff
@@ -606,5 +607,45 @@ private void problemShouldHaveProvidedCodesId(ProblemView problem, String fileId
...
     }
+
+    @Test
+    void GivenOneProblemSaved_WhenArchiveIt_ThenProblemProvidedCodesAndTestcaseIOsShouldBeExist() throws Exception {
+        int problemId = 1;
+        Problem problem = saveProblemAndGet(problemId);
+
+        archiveOrDeleteProblem(problemId);
+
+        assertTrue(problemRepository.findProblemById(problemId).orElseThrow().isArchived());
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java:618]: 這個其他測試案例已經測過了, 所以可以不用留

File: Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java [hunk 7]
```diff
@@ -606,5 +607,45 @@ private void problemShouldHaveProvidedCodesId(ProblemView problem, String fileId
...
+
+        itsShouldBeExist(problem);
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java:620]: 這個命名跟你想驗證的結果對不上

File: Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java [hunk 8]
```diff
@@ -606,5 +607,45 @@ private void problemShouldHaveProvidedCodesId(ProblemView problem, String fileId
...
+        int problemId = 1;
+        Problem problem = saveProblemAndGet(problemId);
+
+        archiveOrDeleteProblem(problemId);
+
+        assertTrue(problemRepository.findProblemById(problemId).orElseThrow().isArchived());
+
+        archiveOrDeleteProblem(problemId);
+
+        itsShouldBeNotFounds(problem);
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java:634]: ditto

File: Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java [hunk 9]
```diff
@@ -606,5 +607,45 @@ private void problemShouldHaveProvidedCodesId(ProblemView problem, String fileId
...
+        List<String> providedCodes = mapToList(problem.getLanguageEnvs().values(),LanguageEnv::getProvidedCodesFileId);
+        assertTrue(problemRepository.isFileExistsById(testcaseIOsFileId));
+        providedCodes.forEach(id -> assertTrue(problemRepository.isFileExistsById(id)));
+    }
+
+    private void itsShouldBeNotFounds(Problem problem) {
+        String testcaseIOsFileId = problem.getTestcaseIOsFileId();
+        List<String> providedCodes = mapToList(problem.getLanguageEnvs().values(),LanguageEnv::getProvidedCodesFileId);
+        assertFalse(problemRepository.isFileExistsById(testcaseIOsFileId));
+        providedCodes.forEach(id -> assertFalse(problemRepository.isFileExistsById(id)));
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java:648]:
```Java
        List fileIds = new LinkedList<>();
        List<String> providedCodes = mapToList(problem.getLanguageEnvs().values(),LanguageEnv::getProvidedCodesFileId);
        fileIds.addAll(providedCodes);
        fileIds.add(problem.getTestcaseIOsFileId());
        fileIds.foreach(id->assertTrue(!problemRepository.existsFile(id)));
```

**Review by Wally5077 — CHANGES_REQUESTED** [review]
Review done

**Review by Wally5077 — APPROVED** [review]
LGTM

File: Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java [hunk 10]
```diff
@@ -167,7 +170,14 @@ public void patchProblem(int problemId, PatchProblemParams params) {
     @Override
     public boolean problemExists(int problemId) {
         return mongoTemplate.exists(query(where("_id").is(problemId)), ProblemData.class);
+    }
 
+    @Override
+    public boolean existsFile(String fileId) {
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java:176]: should be a private method

`ProblemRepository` has nothing to do with the file

File: Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java [hunk 11]
```diff
@@ -167,7 +170,14 @@ public void patchProblem(int problemId, PatchProblemParams params) {
...
+        return ofNullable(gridFsTemplate.findOne(new Query(where("_id").is(fileId))))
+                .map(gridFsTemplate::getResource)
+                .map(GridFsResource::exists)
+                .orElse(false);
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java:180]: delete this method

move 
```java
gridFsTemplate.findOne(new Query(where("_id").is(fileId))))               
 .map(gridFsTemplate::getResource)                
.map(GridFsResource::exists)               
 .orElse(false)
```
to the Integration test

File: Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java [hunk 12]
```diff
@@ -177,24 +187,41 @@ public boolean problemExists(int problemId) {
...
+        deleteFiles(problem);
     }
 
     @Override
     public void deleteAll() {
+        findAll().forEach(this::deleteFiles);
         mongoTemplate.dropCollection(ProblemData.class);
     }
 
+    private void deleteFiles(Problem problem) {
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java:208]: rename to `deleteProvidedCodesAndTestcaseIOs(Problem problem)`

File: Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java [hunk 13]
```diff
@@ -177,24 +187,41 @@ public boolean problemExists(int problemId) {
...
+        List<String> fileIds = new LinkedList<>();
+        var languageEnvs = problem.getLanguageEnvs();
+        if (languageEnvs != null) {
+            List<String> providedCodesFileIds = mapToList(languageEnvs.values(), LanguageEnv::getProvidedCodesFileId);
+            fileIds.addAll(providedCodesFileIds);
+        }
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java:214]: the `languageEnvs` can never be null

in order to ensure this
modify Problem's `getLanguageEnvs` as follows
```java
public Map<String, LanguageEnv> getLanguageEnvs() {
        return requireNonNullElse(languageEnvs, emptyMap());;
}
```

File: Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java [hunk 14]
```diff
@@ -606,5 +607,41 @@ private void problemShouldHaveProvidedCodesId(ProblemView problem, String fileId
...
     }
+
+    @Test
+    void GivenOneProblemSaved_WhenArchiveIt_ThenProblemProvidedCodesAndTestcaseIOsShouldBeExist() throws Exception {
+        int problemId = 1;
+        Problem problem = saveProblemAndGet(problemId);
+
+        archiveOrDeleteProblem(problemId);
+
+        problemProvidedCodesAndTestcaseIOsShouldBeExist(problem);
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java:618]: delete this testcase

File: Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java [hunk 15]
```diff
@@ -606,5 +607,41 @@ private void problemShouldHaveProvidedCodesId(ProblemView problem, String fileId
...
+
+        problemProvidedCodesAndTestcaseIOsShouldBeDeleted(problem);
+    }
+
+    private void problemProvidedCodesAndTestcaseIOsShouldBeExist(Problem problem) {
+        List<String> providedCodes = mapToList(problem.getLanguageEnvs().values(),LanguageEnv::getProvidedCodesFileId);
+        List<String> fileIds = new LinkedList<>(providedCodes);
+        fileIds.add(problem.getTestcaseIOsFileId());
+        fileIds.forEach(fileId -> assertTrue(problemRepository.existsFile(fileId)));
+    }
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java:638]: remove this method

File: Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java [hunk 16]
```diff
@@ -606,5 +607,41 @@ private void problemShouldHaveProvidedCodesId(ProblemView problem, String fileId
...
+
+    private void problemProvidedCodesAndTestcaseIOsShouldBeDeleted(Problem problem) {
+        List<String> providedCodes = mapToList(problem.getLanguageEnvs().values(),LanguageEnv::getProvidedCodesFileId);
+        List<String> fileIds = new LinkedList<>(providedCodes);
+        fileIds.add(problem.getTestcaseIOsFileId());
+        fileIds.forEach(fileId -> assertFalse(problemRepository.existsFile(fileId)));
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java:644]: directly use `GridFsTemplate`

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
Review done

**Review by Johnny850807 — APPROVED** [review]
ＬＧＴＭ

**Pull request merged**